File: 704. Binary Search/recursive.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        return self.recursive(left, right, target, nums)
    
    def recursive(self, left, right, target, nums):

        if left > right:
            return -1

        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            right = mid - 1
            return self.recursive(left, right, target, nums)
        elif nums[mid] < target:
            left = mid + 1
            return self.recursive(left, right, target, nums)

File: 704. Binary Search/recursive.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {

        int left = 0, right = nums.size() - 1;
        return recursive(nums, target, left, right);
        
    }

    int recursive(vector<int>& nums, int target, int left, int right) {
        if (left > right)
            return -1;

        int mid = (left + right) / 2;
        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] > target) {
            right = mid - 1;
            return recursive(nums, target, left, right);
        } else {
            left = mid + 1;
            return recursive(nums, target, left, right);
        }
    }
};